server/src/main/java/com/linecorp/centraldogma/server/internal/replication/ZooKeeperCommandExecutor.java

```diff
@@ -790,32 +792,70 @@ public void childEvent(CuratorFramework unused, PathChildrenCacheEvent event) th
     }
 
     private SafeCloseable safeLock(Command<?> command) {
+        final long lockTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(lockTimeoutMillis);
         final String executionPath = command.executionPath();
         final InterProcessMutex mtx = mutexMap.computeIfAbsent(
                 executionPath, k -> new InterProcessMutex(curator, absolutePath(LOCK_PATH, k)));
 
         WriteLock writeLock = null;
-        boolean lockTimeout = false;
+        boolean lockAcquired = false;
+        boolean success = false;
+        Throwable cause = null;
         try {
-            // Align with the default request timeout
-            if (!mtx.acquire(10, TimeUnit.SECONDS)) {
-                lockTimeout = true;
-                throw new ReplicationException(
-                        "Failed to acquire a lock for " + executionPath + " in 10 seconds");
-            }
-            if (command instanceof NormalizingPushCommand) {
-                writeLock = acquireWriteLock((NormalizingPushCommand) command);
-            } else if (command instanceof RemoveRepositoryCommand) {
-                clearWriteQuota((RemoveRepositoryCommand) command);
+            // Retry up to 1 minute, to minimize the chance of going read-only.
+            long remainingTimeNanos = lockTimeoutNanos;
+            final long deadlineNanos = System.nanoTime() + remainingTimeNanos;
+            for (;;) {
+                try {
+                    if (mtx.acquire(remainingTimeNanos, TimeUnit.NANOSECONDS)) {
+                        lockAcquired = true;
+                        break;
+                    }
+                } catch (NullPointerException e) {
+                    // We're not sure why this happens, but we're retrying to recover from it.
+                    logger.warn("Unexpected NPE from Curator while acquiring a lock for {} (command: {}):",
+                                executionPath, command, e);
+                }
+
+                // Sleep a little bit to avoid high CPU usage.
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException ignored) {
+                    // Ignore
+                }
+
+                remainingTimeNanos = deadlineNanos - System.nanoTime();
+                if (remainingTimeNanos <= 0) {
+                    // Timed out.
+                    break;
+                }
```

Comment:
nit) What do you think of sleeping after checking the remaining timeout?
```suggestion
                remainingTimeNanos = deadlineNanos - System.nanoTime();
                if (remainingTimeNanos <= 0) {
                    // Timed out.
                    break;
                }

                // Sleep a little bit to avoid high CPU usage.
                try {
                    Thread.sleep(100);
                } catch (InterruptedException ignored) {
                    // Ignore
                }
```